lib/ClangImporter/ClangDerivedConformances.cpp

```diff
@@ -1364,3 +1364,218 @@ void swift::conformToCxxSpanIfNeeded(ClangImporter::Implementation &impl,
+      lhsParamId, lhsTy, lhsTyInfo, clang::StorageClass::SC_None,
+      /*DefArg*/ nullptr);
+  auto lhsParamRefExpr = new (clangCtx) clang::DeclRefExpr(
+      clangCtx, lhsParamDecl, false, lhsTy, clang::ExprValueKind::VK_LValue,
+      clang::SourceLocation());
+
+  auto rhsParamId = nullptr;
+  auto rhsTyInfo = clangCtx.getTrivialTypeSourceInfo(rhsTy);
+  auto rhsParamDecl = clang::ParmVarDecl::Create(
+      clangCtx, opDecl, clang::SourceLocation(), clang::SourceLocation(),
+      rhsParamId, rhsTy, rhsTyInfo, clang::StorageClass::SC_None, nullptr);
+  auto rhsParamRefExpr = new (clangCtx) clang::DeclRefExpr(
+      clangCtx, rhsParamDecl, false, rhsTy, clang::ExprValueKind::VK_LValue,
+      clang::SourceLocation());
+
+  opDecl->setParams({lhsParamDecl, rhsParamDecl});
+
+  // looking for the constructor for the function object
+  clang::CXXConstructorDecl *ctorDecl = nullptr;
+  for (clang::CXXConstructorDecl *candidate : funcDecl->ctors()) {
+    if (candidate->param_size() == 0) {
+      ctorDecl = candidate;
+      break;
+    }
+  }
+
+  if (!ctorDecl)
+    return false;
+
+  clang::QualType funcTy = clangCtx.getRecordType(funcDecl);
+  clang::SourceLocation funcLoc = funcDecl->getLocation();
+
+  clang::ExprResult synthCtorExprResult = clangSema.BuildCXXConstructExpr(
+      funcLoc, funcTy, ctorDecl,
+      /*Elidable=*/false, {},
+      /*HadMultipleCandidates=*/false,
+      /*IsListInitialization=*/false,
+      /*IsStdInitListInitialization=*/false,
+      /*RequiresZeroInit=*/false, clang::CXXConstructionKind::Complete,
+      clang::SourceRange(funcLoc, funcLoc));
+
+  ASSERT(!synthCtorExprResult.isInvalid() &&
+         "Unable to synthesize constructor expression for std::equal_to");
+  clang::Expr *synthCtorExpr = synthCtorExprResult.get();
+
+  SmallVector<clang::Expr *, 2> args = {lhsParamRefExpr, rhsParamRefExpr};
+  auto call =
+      clangSema.BuildCallExpr(nullptr, synthCtorExpr, clang::SourceLocation(),
+                              args, clang::SourceLocation());
+
+  if (!call.isUsable())
+    return false;
+
+  auto opBody = clang::ReturnStmt::Create(clangCtx, clang::SourceLocation(),
+                                          call.get(), nullptr);
+  opDecl->setBody(opBody);
+
+  impl.synthesizedAndAlwaysVisibleDecls.insert(opDecl);
+  auto lookupTable1 = impl.findLookupTable(classDecl);
+  addEntryToLookupTable(*lookupTable1, opDecl, impl.getNameImporter());
+  auto owningModule = impl.getClangOwningModule(classDecl);
+  auto lookupTable2 = impl.findLookupTable(owningModule);
+  if (lookupTable1 != lookupTable2)
+    addEntryToLookupTable(*lookupTable2, opDecl, impl.getNameImporter());
+  return true;
+}
+
+static clang::ClassTemplateSpecializationDecl *
+lookupAndSpecializeFunctionObject(ClangImporter::Implementation &impl,
+                                  const clang::CXXRecordDecl *clangDecl,
+                                  std::string name) {
+  ASSERT(clangDecl);
```

Comment:
nit: I don't think this is necessary: this is a `static` helper function whose caller already asserts this.

```suggestion
```